keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1159 @@
+
+Best practice is likely to avoid ApplySets spanning namespaces.  However, sometimes this is unavoidable - particularly when managing cluster-scoped objects - and the “plumbing” tooling cannot enforce this restriction.
+
+Where a GK is known to be part of the ApplySet and is cluster-scoped, we should naturally query for those objects at cluster scope; any permission problems here should be surfaced as errors.  Where we cannot determine the list of GKs for an ApplySet, we may support “discovery”, likely warning that the ApplySet does not define a list of GKs, and then attempt to perform cluster-scoped queries, likely warning if there are insufficient permissions.
+
+>  <<[UNRESOLVED @justinsb @KnVerey]>>
+>
+> Should kubectl ever fall back to this? Will it have a separate subcommand for it?
+>
+> <<[/UNRESOLVED]>>
+
+For GKs that are namespace scoped, we would normally expect those to be part of an ApplySet object in the same namespace.  We define an additional annotation however for cross-namespace ApplySets:  `applyset.k8s.io/additional-namespaces`.  The value of this annotation is a comma-separated list of the names of the namespaces (other than the ApplySet namespace) in which objects are found, for example `default,kube-system,ns1,ns2`.  Note that there is no need to include this specifically for cluster-scoped objects, as those are covered by the group-kind list.  We reserve the empty value.  If this annotation is present, the tooling will query namespace-scoped resources in those namespaces in addition to the namespace of the ApplySet object (if any).  If this annotation is not present on a namespace-scoped ApplySet parent object, the tooling will query namespace-scoped resources only in the same namespace as the ApplySet parent object.  If the annotation is not present on a cluster-scoped ApplySet parent object, the tooling will not query namespace-scoped resources at all (kubectl will output an error if given namespace-scoped GKs in this case).
+
+As with `applyset.k8s.io/contains-group-kinds`, this list of namespaces must be sorted alphabetically, and should be minimal (ideally other than during apply and prune operations).
+
+As cross-namespace ApplySets are not particularly encouraged, we do not currently optimize this further.  In particular, we do not specify the GKs per namespace.  We can add more annotations in future should the need arise.
+
+### Objects with owner references
+
+>  <<[UNRESOLVED @justinsb @KnVerey]>>
+>
+> If an object in the set we retrieve for pruning
+> has owner references, what should we do? This would be
+> somewhat unexpected; it implies that kubectl apply doesn't
+> really own the object, apparently.
+> We could consider this an error, or perhaps warn
```

Comment:
If the owner reference is to the "parent" applyset object, then I suggest we should ignore that particular case.  It feels very reasonable for a third party tool to create objects with ownerReferences, but still mark them with our labels/annotations for compatibility.

If the owner reference is to something else, I propose that in alpha we should make that an error, to force the discussion.  I can see us eventually making it a warning (or ignored if we see that the parent is also in the applyset), but I think that we should use the alpha to try to discover use-cases here.